store/postgres/src/entities.rs

```diff
@@ -420,7 +420,7 @@ impl SplitTable {
 
             Ok(diesel::update(target)
                 .set((
-                    entities::data.eq(entities::data.merge(&data)),
+                    entities::data.eq(&data),
```

Comment:
This is not right: the `Connection::update` method is also used from `Store::update_operation` where it is necessary that we merge data in the database (the whole point of adding an `Update` operation was to avoid roundtripping data just to merge it, especially for block pointer updates).

The correct way to do this would be to restore the `merge` here and have `apply_set_operation` pass in the `data` that is passed to it. At some point (when we switch history logging to all be in Rust) we can use what we learn from collecting history to distinguish between insert and update. Or we could change `apply_set_operation` now to just check whether the entity exists, and not merge there.

Reply:
I can pass along the `data` and put `entities:data.merge(&data)` back in. But this is a change in behavior. Before we were using

https://github.com/graphprotocol/graph-node/blob/master/store/postgres/src/store.rs#L443-L450

to apply the operation on the Rust side _plus_ the `entities::data.merge`, now we'd just be using `entities::data.merge`.

The difference in behavior is that `operation.apply` merged the existing and new entity data and `Null` values caused fields to be removed. With `entities::data.merge`, they would be preserved but set to `Null`. We can do that. It doesn't change the GraphQL behavior, but direct store access that differentiates between absent fields and `Null` fields would be affected.

If we're lucky, that's just the tests. 😉

Reply:
Yeah, I looked at this a while ago, and it was the reason why I didn't change to `merge` everywhere. I have a half-done PR to do key removal in the DB too (you can write `data - 'key'` in pg to do that) but couldn't get the Diesel mechanics behind it to typecheck. I can have another go at that. In the meantime, we could add a `updated_data: Option<serde_json::Value>` to `Table::update` and merge when that is `None`, and just set when it is `Some`. I'd somehow feel better about that approach than changing the tests, because I am not sure what other issues this will cause down the line. I assume it's rare that mappings remove keys, which means we'd probably run into issues there at the worst possible moment. I can try and come up with a better way to do this once this PR is merged.

Reply:
@lutter Updated again, let me know what you think!